### ISSUE TITLE
element 0 of tensors does not have a grad_fn 

### ISSUE THREAD

**Issue by lepoeme20** [PyTorchLightning/pytorch-lightning#2427]
#### What is your question?
Hi,
I got RuntimeError
element 0 of tensors does not have a grad_fn in **TEST phase**

#### Code
```
def test_step(self, batch, batch_nb):
    imgs, labels = batch
    self.last_imgs = imgs

    # get gradients
    self.classifier.eval()
    copied_input = imgs.clone()
    copied_input.requires_grad_(True)
    output = self.classifier(copied_input)
    loss = self.criterion_ce(output, labels)
    gradients = torch.autograd.grad(loss, copied_input)[0]
    self.last_grad = gradients

def criterion_ce(self, y_hat, y):
    return F.cross_entropy(y_hat, y)
```
#### What have you tried?
Of course, the above code snippet works well in the training phase.
However, grad_fn of output and loss is None in the test/val phase (In the training phase, there are AddmmBackward and NllLossBackward, respectively).

How can I get the input tensor gradients in the Test/Val phase?

#### What's your environment?

 - OS: Linux
 - conda

**Comment by williamFalcon** [comment]
you can enable gradients again inside the beginning of the test loop.

```
def test_step(...):
   self.train()
```

But generally, test/eval is done without gradients.

**Comment by lepoeme20** [comment]
@williamFalcon 
Thank you for your answer.

I understand what you are saying. But that doesn't solve the problem.

In my research, input gradient is also required in test phase.
Also, as shown below, my model (self) is a GAN generator, not a classifier.

```
class Generator(nn.Module):
    '''

class Discriminator(nn.Module):
    ''''

class Classifier(nn.Module):
    model = torch.load(checkpoint_dir)

    for params in clf.parameters():
         params.requires_grad = False

    return model

class MyModel(pl.LightningModule):
    self.classfier = '''

'''

def test_step(self, batch, batch_nb):
    imgs, labels = batch
    self.last_imgs = imgs

    # get gradients
    self.classifier.eval()
    copied_input = imgs.clone()
    copied_input.requires_grad_(True)
    output = self.classifier(copied_input)
    loss = self.criterion_ce(output, labels)
    gradients = torch.autograd.grad(loss, copied_input)[0]
    self.last_grad = gradients

def criterion_ce(self, y_hat, y):
    return F.cross_entropy(y_hat, y)
```

Adding self.train() or self.classifier.train() to the beginning of test_step() like your word does not solve my problem.

**Comment by nilsleh** [comment]
@lepoeme20  I just ran into a similar issue, where I require gradients during the test phase, and similar to what you were saying 
`self.train()` or `self.classifier.train()` did not do anything. I know this is years later, however, what did solve it was to also call `torch.set_grad_enabled(True)` at the beginning of test_step. I believe under the hood, pytorch lightning calls this context manager and does not allow gradient computations that way, so calling `.train()` does not change anything, but with `torch.set_grad_enabled(True)` I get gradients.

**Comment by mvinyard** [comment]
@nilsleh This is interesting. I am in a similar boat: I have a task that requires gradients during the test phase. This tip you mention of including `torch.set_grad_enabled(True)` is really useful and works for the validation step, for me. However, I'm still running into this issue during the `LitModel.test_step()` step. If you found any other tricks, please drop them here! Nonetheless, appreciate your above comment!

**Comment by mvinyard** [comment]
@lepoeme20 @nilsleh @williamFalcon 

As an update, this workaround seemed to do the trick (copying a [suggestion](https://github.com/Lightning-AI/lightning/pull/770#issuecomment-750855675) from @williamFalcon in #770):

```python
def evaluate_retain_grad(self):
    self.trainer.num_sanity_val_steps = -1
    self.trainer.fit(self, self._dataset['train'], self._dataset['test'])
```

Where `evaluate_retain_grad` is a method of the wrapper class for the `LightningModule` and `trainer` is `Trainer()` saved as an attribute of this wrapper, as is `dataset` being a `torch.utils.data.dataloader.DataLoader`.

**Comment by alfieroddan** [comment]
I also am in need of gradients at test step for my research!

@williamFalcon Makes sense that it isn't enabled by default but why should it behave different to validation (I imagine something to do with freezing the optimizer param group)? I couldn't find the code in the core loops that causes this behaviour.

To be clear of my issue is during `test_step` my logits do **not** `require_grad` or have `grad_fn` needed to do a `.backward()`. I have done `torch.set_grad_enabled(True)` and `for p in self.model.parameters(): p.requires_grad =True;` in `test_step` and a callback before doing a forward pass of the model to rectify with no help (I can provide some example code if helpful but others have done above).

If it's helpful I could open another issue?

@mvinyard Thanks for the work around, I ended up just replacing `trainer.test` with my own  loop for logging purposes.

**Comment by stases** [comment]
This is still not solved, could you please reopen?
